src/client/clientmap.cpp

```diff
@@ -742,49 +776,68 @@ void ClientMap::renderMapShadows(video::IVideoDriver *driver,
 					video::SMaterial &mat = buf->getMaterial();
 					auto rnd = driver->getMaterialRenderer(mat.MaterialType);
 					bool transparent = rnd && rnd->isTransparent();
-					if (transparent == is_transparent_pass)
-						drawbufs.add(buf, block_pos, layer);
+					if (!transparent)
+						grouped_buffers.add(buf, block_pos, layer);
 				}
 			}
 		}
 	}
 
+	// Capture draw order for all solid meshes
+	for (auto &lists : grouped_buffers.lists) {
+		for (MeshBufList &list : lists) {
+			// iterate in reverse to draw closest blocks first
+			for (auto it = list.bufs.rbegin(); it != list.bufs.rend(); ++it) {
+				draw_order.emplace_back(it->first, it->second, it != list.bufs.rbegin());
+			}
```

Comment:
Could use a reserve call *before* the loop, this probably works:
`draw_order.reserve(draw_order.size() + list.bufs.size());`

Reply:
I decided to preallocate draw_order once, since there are a lot of buffers, and moving memory feels more expensive.